SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ComputeSideEffects.swift

```diff
@@ -142,6 +142,24 @@ private struct CollectedEffects {
         handleApply(pa)
         checkedIfDeinitBarrier = true
       }
+      // In addition to the effects of the apply, also consider the
+      // effects of the capture, which reads the captured value in
+      // order to move it into the context. This only applies to
+      // addressible values, because capturing does not dereference
+      // any class objects.
+      //
+      // Ignore captures for on-stack partial applies. They only
+      // bitwise-move or capture by address, so the call to
+      // handleApply above is sufficient. And, if they are not applied
+      // in this function, then they are never applied.
+      if !pa.isOnStack {
+        // the callee and its arguments are all captured...
+        for operand in pa.operands {
+          if operand.value.type.isAddress {
```

Comment:
@atrick For the record: the callee (operand 0) never has an address type. So you could write `for argument in pa.arguments`. It's just cosmetic, but would make it clear that only arguments can contribute to the read-effects.

Reply:
@eeckstein I didn't think the callee could be an address but also didn't see anything in the SILVerifier prohibiting it. The closure result must be an object of course, but it says nothing about the captured callee. I suppose we never reabstract over function types in a way that requires an address, so I'll just add an assert to SILVerifier and fix this code as you suggested.